Doc: mention possibility to reduce TTL for cached results when uploading blocks (#5410)
grafana/mimir: docs/sources/mimir/configure/configure-tsdb-block-upload.md

  compactor_block_upload_enabled: true
```

### Validation of blocks

> Before upload of block data starts, Grafana Mimir performs following checks on `meta.json` file:

Reviewer: ```suggestion
Before uploading block data starts, Grafana Mimir performs the following checks on the `meta.json` file:
```

> * Only TSDB "v1" blocks are supported. This is format used by Prometheus v2, Grafana Mimir or Thanos.

Reviewer: ```suggestion
* Only TSDB "v1" blocks are supported. This is the format used by Prometheus v2, Grafana Mimir, and Thanos.
```

> * Blocks with invalid MinTime or MaxTime are rejected (negative values or MaxTime < MinTime)

Reviewer: ```suggestion
* Blocks with invalid MinTime or MaxTime are rejected (negative values or MaxTime < MinTime).
```

> * Blocks where MinTime or MaxTime is in the future are rejected

Reviewer: ```suggestion
* Blocks where MinTime or MaxTime is in the future are rejected.
```

> * Blocks that are outside of retention period are rejected

Reviewer: ```suggestion
* Blocks that are outside of the retention period are rejected.
```

> * Blocks covering time range larger than maximum compaction range (`-compactor.block-ranges` option, maximum defaults to 24h) are rejected

Reviewer: ```suggestion
* Blocks covering a time range larger than the maximum compaction range (`-compactor.block-ranges` option, maximum defaults to 24h) are rejected.
```

> * Blocks which cross the boundary of maximum compaction range are rejected. For example if largest compaction range is 24h, blocks that start before midnight and finish after midnight would be rejected.

Reviewer: ```suggestion
* Blocks that cross the boundary of the maximum compaction range are rejected. For example, if the largest compaction range is 24 hours, blocks that start before midnight and finish after midnight are rejected.
```

> * Blocks with Thanos downsampling configuration are rejected

Reviewer: ```suggestion
* Blocks with Thanos downsampling configuration are rejected.
```

> * Blocks that are bigger than `compactor_block_upload_max_block_size_bytes` (per-tenant override) are rejected.

Reviewer: ```suggestion
* Blocks that are larger than the `compactor_block_upload_max_block_size_bytes` (per-tenant override) setting are rejected.
```

> * Blocks with "external labels" (Thanos feature) are rejected. (Some Mimir-specific labels are allowed)

Reviewer: ```suggestion
* Blocks with "external labels" (Thanos feature) are rejected. Some Mimir-specific labels are allowed.
```

> After block index and chunks are uploaded, Grafana Mimir performs additional block validation of block index and chunks to verify that blocks are well-formed, and they cannot possibly cause problems for Mimir operation.

Reviewer: ```suggestion
After you upload the block index and chunks, Grafana Mimir performs additional block validation to verify that blocks are well-formed and that they won't cause problems for Mimir's operation.
```

> These "full block" validations can be disabled via `compactor_block_upload_validation_enabled` per-tenant override.

Reviewer: ```suggestion
You can disable these "full block" validations through the `compactor_block_upload_validation_enabled` per-tenant override.
```

> To disable chunks validation only but keep index-validation `compactor_block_upload_verify_chunks` per-tenant override can be used instead.

Reviewer: ```suggestion
To disable chunks validation while keeping index-validation, use the `compactor_block_upload_verify_chunks` per-tenant override instead.
```

### The results-cache needs flushing

> Grafana Mimir caches query results for queries that don't require the most recent 10 minutes of data.

Reviewer 2: This is not entirely correct. Queries requirement the last 10m are still cached, but we don't cache the last 10m samples. I suggest to improve it.

Author: WDYT:

> Grafana Mimir caches query results that are 10 minutes old.

(This is just introduction sentence to explain the problem with caching and backfill)

Reviewer 2: WDYT about:

> Grafana Mimir caches samples older than 10 minute (configurable via `-query-frontend.max-cache-freshness`) in the range query results.

Now, that 10m is even wrong. The default is 1m 😱 . Going to open a PR to change the default in Mimir, cause I think we've always communicated 10m to everyone.

Author: I still see default being 10m :thinking:

After uploading blocks however, queries may return different results – because new data was uploaded.
This means that cached results may be wrong.

> To fix the cache results, Mimir operator can manually flush the results cache.
> Possible alternative is to decrease time-to-live period for cache results from default 7 days to shorter period, for example 6 hours, by using `-query-frontend.results-cache-ttl` command line option (or per tenant).
> This will guarantee that query results will use backfilled data at most after this period.  

Reviewer 3: Language nits. I think that qualifying the alternative as possible is unnecessary, to me at least it makes it sound as if it's just a "possible" alternative and not a certain one.
```suggestion
To fix the cache results, the Mimir operator can manually flush the results cache.
An alternative is to decrease the time-to-live period for cache results from the default 7 days to a shorter period, for example 6 hours, by using `-query-frontend.results-cache-ttl` command line option (or per tenant).
This will guarantee that query results will use backfilled data at most after this period.
```

Reviewer 4: ```suggestion
To fix the cache results, the Mimir operator can manually flush the results cache.
An alternative is to decrease the time-to-live period for cache results from the default `7` days to a shorter period, for example 6 hours, by using the `-query-frontend.results-cache-ttl` command-line option (or per tenant).
This guarantees that query results will use backfilled data at most after this period.
```
What does "at most" mean here? Does it mean "mostly"?

Author: It means that you need to wait at most 6h after upload finished for uploaded data to appear in query results.
